### PR TITLE
feat: implement allow disruptive action

### PR DESCRIPTION
**Pull request by fzipi**
Signed-off-by: Felipe Zipitria <felipe.zipitria@owasp.org>

Implements the `allow` disruptive action.

### CODE REVIEW

File: actions/allow.go [hunk 1]
```diff
@@ -6,39 +6,44 @@ package actions
...
-			//tx.SkipToPhase = tx.LastPhase +1
-		} else if a.allow == 3 && tx.LastPhase < 3 {
-			//tx.SkipToPhase = 3
-		}
-	*/
+// Evaluate Allow has the following rules:
+//
+// Example: `SecRule REMOTE_ADDR "^192\.168\.1\.100$" "phase:1,id:95,nolog,allow"`
+//
+//   - If used one its own, like in the example above, allow will affect the entire transaction,
```
Comment (jcchavezs) [inline actions/allow.go:37]:
```suggestion
//   - If used on its own, like in the example above, allow will affect the entire transaction,
```
maybe a typo?

File: types/allow.go [hunk 2]
```diff
@@ -0,0 +1,14 @@
+// Copyright 2022 Juan Pablo Tosso and the OWASP Coraza contributors
+// SPDX-License-Identifier: Apache-2.0
+
+package types
+
+// AllowType represents the values that the allow disruptive action can take
+type AllowType int
```
Comment (anuraaga) [inline types/allow.go:7]: Looks like this isn't exposed in a public API so we should keep the enum internal too

Comment (fzipi) [reply types/allow.go:7]: You are right!

File: actions/allow.go (hunk 1 above)
Comment (fzipi) [reply actions/allow.go:37]: Definitely.

File: types/allow.go (hunk 2 above)
Comment (fzipi) [reply types/allow.go:7]: I'm gonna piggyback on @jptosso changes later today.

**Comment by jcchavezs** [comment]
Any movement on this?

**Comment by fzipi** [comment]
I was going to piggyback on the internal movement that @jptosso was doing, and create some helpers around. But looks like that one never got pushed...

**Comment by jptosso** [comment]
moving to internal was canceled because wasm connector was using many things we were going to unexport.

**Comment by fzipi** [comment]
Then I'll review where I was on this and push it again.

**Comment by jcchavezs** [comment]
Friendly ping @fzipi

**Comment by fzipi** [comment]
Thanks @jcchavezs ! More friendly pong!

**Comment by jptosso** [comment]
doesn't look good to me

**Comment by fzipi** [comment]
This one is ready to merge now.

**Comment by jcchavezs** [comment]
PTAL @anuraaga

**Comment by jptosso** [comment]
PTAL @anuraaga

File: internal/corazawaf/waf.go [hunk 3]
```diff
@@ -161,6 +161,7 @@ func (w *WAF) newTransactionWithID(id string) *Transaction {
 	tx.ruleRemoveByID = nil
 	tx.ruleRemoveTargetByID = map[int][]ruleVariableParams{}
 	tx.Skip = 0
+	tx.AllowType = 0
```
Comment (jptosso) [inline internal/corazawaf/waf.go:164]: Please use the constant

Comment (fzipi) [reply internal/corazawaf/waf.go:164]: Fixed, thanks.

**Review by anuraaga — APPROVED** [review]

**Pull request merged**